gpflow/experimental/check_shapes/base_types.py

```diff
@@ -0,0 +1,24 @@
+# Copyright 2022 The GPflow Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Definitions of commonly used types.
+"""
+from typing import Any, Callable, TypeVar, Union
+
+import numpy as np
+import tensorflow as tf
+
+C = TypeVar("C", bound=Callable[..., Any])
+
+Shaped = Union[np.ndarray, tf.Tensor]
```

Comment:
Seems like a duplicate of `TensorType` or `TensorLike`: https://github.com/GPflow/GPflow/blob/c9f88c1f10cc869fcd3556948d3a9a188ddb3246/gpflow/base.py#L34?

Reply:
> Seems like a duplicate of `TensorType` or `TensorLike`:
> 
> https://github.com/GPflow/GPflow/blob/c9f88c1f10cc869fcd3556948d3a9a188ddb3246/gpflow/base.py#L34
> 
> ?

Two things:

1. I'd kind of like for the check_shapes stuff not to depend on the rest of GPflow, so that it's easy to spin into a separate repository later.
2. I don't think the list `[np.ndarray, tf.Tensor]` is the same as the list `[tf.Tensor, tf.Variable, "Parameter"]`. Though maybe both lists should be extended. If we weren't using an ancient version of TensorFlow there is also: https://www.tensorflow.org/api_docs/python/tf/types/experimental/TensorLike

Reply:
I've replaced `Shaped` with `TensorType`.